File: SummerCal/Services/EarningsService.swift
import Foundation

final class EarningsService {
    private let calendar = Calendar.current

    func monthlyEarnings(month: Date, entries: [IncomeEntry]) -> Double {
        let monthly = entries.filter { calendar.isDate($0.date, equalTo: month, toGranularity: .month) }
        return monthly.reduce(0) { $0 + $1.amount }
    }

    func monthlyWorkEarnings(month: Date, sessions: [WorkSession]) -> Double {
        let monthly = sessions.filter { calendar.isDate($0.date, equalTo: month, toGranularity: .month) }
        return monthly.reduce(0) { $0 + $1.totalEarned }
    }

    func monthlyTotal(month: Date, entries: [IncomeEntry], sessions: [WorkSession]) -> Double {
        monthlyEarnings(month: month, entries: entries) + monthlyWorkEarnings(month: month, sessions: sessions)
    }

    func goalProgress(total: Double, goal: Double) -> Double {
        guard goal > 0 else { return 0 }
        return min(total / goal, 1.0)
    }

    func dailyAverage(month: Date, total: Double) -> Double {
        guard let range = calendar.range(of: .day, in: .month, for: month) else { return 0 }
        return total / Double(range.count)
    }
}

File: SummerCalExpo/src/services/EarningsService.ts
import { IncomeEntry, WorkSession } from '../models';

export class EarningsService {
  monthlyIncomeEntries(date: Date, entries: IncomeEntry[]): number {
    const target = new Date(date);
    return entries
      .filter(e => new Date(e.date).getMonth() === target.getMonth() && new Date(e.date).getFullYear() === target.getFullYear())
      .reduce((s, e) => s + e.amount, 0);
  }

  monthlyWorkEarnings(date: Date, sessions: WorkSession[]): number {
    const target = new Date(date);
    return sessions
      .filter(s => new Date(s.date).getMonth() === target.getMonth() && new Date(s.date).getFullYear() === target.getFullYear())
      .reduce((s, e) => s + e.totalEarned, 0);
  }

  monthlyTotal(date: Date, entries: IncomeEntry[], sessions: WorkSession[]): number {
    return this.monthlyIncomeEntries(date, entries) + this.monthlyWorkEarnings(date, sessions);
  }

  goalProgress(total: number, goal: number): number {
    if (goal <= 0) return 0;
    return Math.min(total / goal, 1);
  }

  dailyAverage(date: Date, total: number): number {
    const daysInMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0).getDate();
    return total / daysInMonth;
  }
}
